src/pk.c

```diff
@@ -6959,793 +7110,1199 @@ WOLFSSL_BIGNUM* wolfSSL_DH_8192_prime(WOLFSSL_BIGNUM* bn)
+        dh->q = q;
+        dh->g = g;
+        /* External DH key parameters were set. */
+        dh->exSet = 1;
+
+        /* Set internal/wolfSSL DH key as well. */
+        if (SetDhInternal(dh) != 1) {
+            WOLFSSL_ERROR_MSG("Unable to set internal DH key");
+            /* Don't keep parameters on failure. */
+            dh->p = NULL;
+            dh->q = NULL;
+            dh->g = NULL;
+            /* Internal and external DH key not set. */
+            dh->inSet = 0;
+            dh->exSet = 0;
+            ret = 0;
         }
     }
 
-end:
-    if (memAlloced) XFREE(mem, NULL, DYNAMIC_TYPE_PEM);
-    if (der != NULL) FreeDer(&der);
-    XFREE(p, NULL, DYNAMIC_TYPE_PUBLIC_KEY);
-    XFREE(g, NULL, DYNAMIC_TYPE_PUBLIC_KEY);
-    return localDh;
-#else
-    (void)bio;
-    (void)x;
-    (void)cb;
-    (void)u;
-    return NULL;
-#endif
+    return ret;
 }
 
-#ifndef NO_FILESYSTEM
-/* Reads DH parameters from a file pointer into WOLFSSL_DH structure.
+/* Set the length of the DH private key in bits.
  *
- * fp  file pointer to read DH parameter file from
- * x   output WOLFSSL_DH to be created and populated from fp
- * cb  password callback, to be used to decrypt encrypted DH parameters PEM
- * u   context pointer to user-defined data to be received back in password cb
+ * Length field is checked at generation.
  *
- * Returns new WOLFSSL_DH structure pointer on success, NULL on failure. */
-WOLFSSL_DH *wolfSSL_PEM_read_DHparams(XFILE fp, WOLFSSL_DH **x,
-        wc_pem_password_cb *cb, void *u)
+ * @param [in, out] dh   DH key to set.
+ * @param [in]      len  Length of DH private key in bytes.
+ * @return  0 on failure.
+ * @return  1 on success.
+ */
+int wolfSSL_DH_set_length(WOLFSSL_DH *dh, long len)
 {
-    WOLFSSL_BIO* fbio = NULL;
-    WOLFSSL_DH* dh = NULL;
+    int ret = 1;
 
-    if (fp == NULL) {
-        WOLFSSL_MSG("DH parameter file cannot be NULL");
-        return NULL;
+    WOLFSSL_ENTER("wolfSSL_DH_set_length");
+
+    /* Validate parameter. */
+    if (dh == NULL) {
+        WOLFSSL_ERROR_MSG("Bad function arguments");
+        ret = 0;
+    }
+    else {
+        /* Store length. */
+        dh->length = (int)len;
     }
 
-    fbio = wolfSSL_BIO_new(wolfSSL_BIO_s_file());
-    if (fbio == NULL) {
-        WOLFSSL_MSG("Unable to create file BIO to process DH PEM");
-        return NULL;
+    return ret;
+}
+#endif /* OPENSSL_ALL || (v1.1.0 or later) */
+#endif
+
+/* Get the public and private keys requested.
+ *
+ * @param [in]  dh         DH key to get keys from.
+ * @param [out] pub_key    Pointer to return public key in. May be NULL.
+ * @param [out] priv_key   Pointer to return private key in. May be NULL.
+ */
+void wolfSSL_DH_get0_key(const WOLFSSL_DH *dh, const WOLFSSL_BIGNUM **pub_key,
+    const WOLFSSL_BIGNUM **priv_key)
+{
+    WOLFSSL_ENTER("wolfSSL_DH_get0_key");
+
+    /* Get only when valid DH passed in. */
+    if (dh != NULL) {
+        /* Return public key if available and available. */
```

Comment:
available and available -> required and available

Reply:
Fixed